### DISCUSSION TITLE
Colour shifting approach circle mod

### DISCUSSION THREAD

**Discussion by mk56-spn** [ppy/osu#13195] [Ideas]
(I'm rather new to this sorry if I mess up the request format slightly)

The idea is rather simple  but id say rather intruiging aswell (may have come up before but  i googled and checked git as best as i could to no avail) .

So basically the approach circles (only the actual circle part there would be no circumference closing  in in this concept since it would serve no purpose)  would (or a designated part of them depending on the skin )  start off as a color A ( lets say red  as an example ) the second the circle becomes visible (would depend on the AR obviously and in this sense be no different to traditional approach circles) and end  on another colour B at  the exact moment at which you are meant to click the circle (the transition between these two colours would be the color gradient ) here is an illustration to clarify what I mean :
![heatwave](https://user-images.githubusercontent.com/74463310/116791448-b313f280-aaba-11eb-8f94-734810538250.png)

As to why I think this would be of interest: 
It would offer a different approach to reading without being inherently easier or harder than normal reading (atleast not objectively speaking ) except with hidden which it would nerf as it gives additional info on when to hit which isnt the case for no mod since its simply replacing the circumferences  (and therefore this could be allowed on ranked scores which I dont see why it wouldn't be but thats not relevant to the scope of this proposal) .

As for a proposed name : "Heatwave", simple and I feel it gets the point across pretty well once you see what it does.

Unfortunately I dont know how easy or difficult this would be to implement since my only coding experience is with web programming. that said it doesn't seem like a monumental task.

**Comment by FrameMuse** [comment]
I think it shouldn't be in game by default, but could be possiblly added by a skin change, like customizable skin or just a setting.

**Comment by frenzibyte** [comment]
I would disagree with that, especially since this is changing the way you play the ruleset and is *potentially* increasing the difficulty.

I don't see the problem with having it as a fun mod? or even a difficulty increase mod if decided.

**Reply by mk56-spn** [reply]
I'd opt for fun mod because it doesnt objectively increase difficulty all the time ( neither does hidden but that's another can of worms ). Do you have any idea of how this could be done? 

Because I would give it a shit myself if someone points me in the right direction. 

Ps . Sorry for the ridiculously late answer

**Reply by frenzibyte** [reply]
https://github.com/ppy/osu/blob/f144fcf16ece622198765d5bc5579d12093ecd12/osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs#L506-L524

Hit objects give themselves a colour from the skin combo colours, by assigning it to their bindable `AccentColour`, which subcomponents inside of the hit object would tint themselves from.

What you would want is disabling the hitobject from changing its colour, and transforming the `AccentColour` bindable just like any other mod would do to other parts of the hitobject. (`ModHidden`, `OsuModTransform`, `OsuModSpinIn`, etc..)

Give the [osu!framework wiki](https://github.com/ppy/osu-framework/wiki) a read if you haven't already, the bindables page especially since you'll be dealing with bindables most of the time.

**Reply by mk56-spn** [reply]
Alright will do thank you for the advice !

**Comment by mk56-spn** [comment]
I will revive this soon

**Comment by peppy** [comment]
I'd argue this should probably only affect the approach circle.

**Reply by mk56-spn** [reply]
mm, i think it should override the beatmap colour sections aswell, can we settle on making it a suboption?

**Reply by frenzibyte** [reply]
I was under the thought that the approach circle would hide completely, and that the que to click the circle would be via the current colour of the circle? (click once it reaches green)

**Reply by mk56-spn** [reply]
late answer but yes this is a better option

**Comment by mk56-spn** [comment]
https://github.com/mk56-spn/osu/tree/colour-approach-mod ive kinda reached an impass with the last step of making the colours flash appropritely, since its non functional im posting it here

**Comment by smoogipoo** [comment]
I don't know how I feel about this being a mod rather than a skinning option... That's to say, the distinction between the two becomes very unclear to me if this were to be a mod.

**Reply by peppy** [reply]
Not sure about it being a skin option, since it removes a core gameplay mechanic (combo colouring).

**Reply by smoogipoo** [reply]
Many players play without combo colouring. Wouldn't call it "core".

**Reply by peppy** [reply]
I'll cry in a corner.

**Reply by smoogipoo** [reply]
Regardless of that my main concern is a concern which I've brought up many times in the past, which is that the line between what is a user preference (skin/options) and what is a challenge (mods) is being blurred.
In the same vein I don't think that mania snap colouring should be (and rightly isn't) a mod, because it's a user preference. A player will either play with it always on or always off, not an in-between - a preference. Maybe a preference depending on the skin, e.g. how mania's snap colouring doesn't work if the skin is already coloured.

I was actually fine with the visuals in the PR as opposed to your suggestion of only affecting the approach circle, because I'd see snap colouring (as per mania) work exactly the same way. I could imagine a dropdown in game/skin options named "circle colouring" with options combo/timing/snapping.

**Reply by Gamingzone123** [reply]
id say probably a mod because along with the above, it's quite similar to target practice (kind of a combo of target practice and hidden), probably under fun. I *could* see it as a skinning option, but the higher toggle-ability of a mod makes more sense to me.

**Reply by peppy** [reply]
Keep in mind this isn't snap colouring but rainbow colouring, something very different.

If we decide this should be at a settings or skinning level then that is fine, but it should be well defined before implemented. Since it would probably be a dropdown (if settings) or a property for drawables (if skinning) with options of

- Beatmap provided
- Snap based
- Dynamic rainbow (what this proposal is, need a better name)

**Reply by Moltanicaa** [reply]
Maybe make it part of the Difficulty adjust...?

**Reply by smoogipoo** [reply]
As part of DA mod sounds even worse to me.

**Reply by mk56-spn** [reply]
> Keep in mind this isn't snap colouring but rainbow colouring, something very different.
> 
> If we decide this should be at a settings or skinning level then that is fine, but it should be well defined before implemented. Since it would probably be a dropdown (if settings) or a property for drawables (if skinning) with options of
> 
> * Beatmap provided
> * Snap based
> * Dynamic rainbow (what this proposal is, need a better name)

well its not rainbow colouring, thats just colours ive picked out, fundamentally speaking i think those colours should be up to the end user, stilll i very much understand your concerns peppy im just kinda not sure where to go from here :P
